pkg/controller/persistentvolumelabel/pvlcontroller.go

```diff
@@ -0,0 +1,421 @@
+	// If that gets standardized we can refactor to reduce code duplication
+	spec := aws.KubernetesVolumeID(volume.Spec.AWSElasticBlockStore.VolumeID)
+	labels, err := ebsVolumes.GetVolumeLabels(spec)
+	if err != nil {
+		return nil, err
+	}
+
+	return labels, nil
+}
+
+// getEBSVolumes returns the AWS Volumes interface for ebs
+func (pvlc *PersistentVolumeLabelController) getEBSVolumes() (aws.Volumes, error) {
+	pvlc.mutex.Lock()
+	defer pvlc.mutex.Unlock()
+
+	if pvlc.ebsVolumes == nil {
+		awsCloudProvider, ok := pvlc.cloud.(*aws.Cloud)
+		if !ok {
+			// GetCloudProvider has gone very wrong
+			return nil, fmt.Errorf("error retrieving AWS cloud provider")
+		}
+		pvlc.ebsVolumes = awsCloudProvider
+	}
+	return pvlc.ebsVolumes, nil
+}
+
+func (pvlc *PersistentVolumeLabelController) findGCEPDLabels(volume *v1.PersistentVolume) (map[string]string, error) {
+	// Ignore any volumes that are being provisioned
+	if volume.Spec.GCEPersistentDisk.PDName == vol.ProvisionedVolumeName {
+		return nil, nil
+	}
+
+	provider, err := pvlc.getGCECloudProvider()
+	if err != nil {
+		return nil, err
+	}
+
+	// If the zone is already labeled, honor the hint
+	zone := volume.Labels[kubeletapis.LabelZoneFailureDomain]
+
+	labels, err := provider.GetAutoLabelsForPD(volume.Spec.GCEPersistentDisk.PDName, zone)
+	if err != nil {
+		return nil, err
+	}
+
+	return labels, nil
+}
+
+// getGCECloudProvider returns the GCE cloud provider, for use for querying volume labels
+func (pvlc *PersistentVolumeLabelController) getGCECloudProvider() (*gce.GCECloud, error) {
+	pvlc.mutex.Lock()
+	defer pvlc.mutex.Unlock()
+
+	if pvlc.gceCloudProvider == nil {
+		gceCloudProvider, ok := pvlc.cloud.(*gce.GCECloud)
+		if !ok {
+			// GetCloudProvider has gone very wrong
+			return nil, fmt.Errorf("error retrieving GCE cloud provider")
+		}
+		pvlc.gceCloudProvider = gceCloudProvider
+	}
+	return pvlc.gceCloudProvider, nil
+}
+
+func (pvlc *PersistentVolumeLabelController) updateVolume(volume *v1.PersistentVolume) error {
+	//func (pvlc *PersistentVolumeLabelController) updateVolume(labels map[string]string) error {
+	glog.V(4).Infof("updating PersistentVolume %s", volume.Name)
+	err := clientretry.RetryOnConflict(updatePVLabelsBackoff, func() error {
+		curVol, err := pvlc.kubeClient.Core().PersistentVolumes().Get(volume.Name, metav1.GetOptions{IncludeUninitialized: true})
+		if err != nil {
+			return err
+		}
+		oldData, err := json.Marshal(curVol)
+		if err != nil {
+			return fmt.Errorf("failed to marshal old persistentvolume %#v for persistentvolume %q: %v", curVol, volume.Name, err)
+		}
+
+		objCopy, err := scheme.Scheme.DeepCopy(curVol)
+		if err != nil {
+			return fmt.Errorf("failed to copy persistentvolume object %#v: %v", curVol, err)
+		}
+		newVolume, ok := (objCopy).(*v1.PersistentVolume)
+		if !ok {
+			return fmt.Errorf("failed to cast copy into persistentvolume object %#v: %v", newVolume, err)
+		}
+		newVolume.Labels = volume.Labels
```

Comment:
if all you need/want is a name and a set of labels, then pass just that through as args